File: server/common/protocol/betbatch_message.py
from common.protocol.bet_message import BetMessage
from common.protocol.message import Message


class BetBatchMessage(Message):
    BETBATCH_OP = bytes([0x05])
    
    def __init__(self):
        self.bets = []

    def add_bets(self, bets: list):
        self.bets += bets

    def __eq__(self, __value: object) -> bool:
        try:    
            result = len(self.bets) == len(__value.bets) and all(map(lambda x: x[0] == x[1], zip(self.bets, __value.bets)))
        except:
            result = False
        finally:
            return result
        

    def serialize(self) -> bytes:
        serialized_bets = bytes()
        for bet in self.bets:
            serialized_bets += bet.serialize()
        header = self._build_header(self.BETBATCH_OP, len(serialized_bets))

        return header + serialized_bets
    
    def __deserialize_bet(self, stream: bytes) -> (bytes, bool):
        betsize = self._get_message_length(stream)
        if betsize < 0:
            return bytes(), False
        new_bet = BetMessage()
        ok = new_bet.deserialize(stream[:betsize])
        if not ok:
            return bytes(), False
        self.bets.append(new_bet)
        return stream[betsize:], ok
    
    def deserialize(self, stream: bytes) -> bool:
        if not self._check_header(stream, self.BETBATCH_OP):
            return False
        ok = True
        stream = stream[self.HEADER_SIZE:]
        while len(stream) > 0 and ok:
            stream, ok = self.__deserialize_bet(stream)
            

        return ok
    

File: client/common/protocol/betbatch_message.go
package protocol

type BetBatch struct {
	Bets []Bet
}

func (betb *BetBatch) Serialize() []byte {
	serializedBets := make([]byte, 0)

	for _, bet := range betb.Bets {
		serializedBets = append(serializedBets, bet.Serialize()...)
	}

	header := betb.makeHeader(len(serializedBets))

	return append(header, serializedBets...)
}

func (betb *BetBatch) Deserialize(stream []byte) (err error) {

	err = checkHeader(stream, BETBATCH_OP)

	if err != nil {
		return
	}

	stream = stream[HEADER_SIZE:]

	for len(stream) > 0 && err == nil {
		err = betb.deserializeBetMessage(&stream)
	}
	return
}

func (betb *BetBatch) ShouldAck() bool {
	return true
}

func (betb BetBatch) makeHeader(betsLength int) []byte {
	header := []byte{BETBATCH_OP}
	return buildHeader(header, betsLength)
}

func (betb BetBatch) getBetStreamSize(stream []byte) (int, error) {
	return GetMessageLength(stream)
}

func (betb *BetBatch) deserializeBetMessage(stream *[]byte) (err error) {

	betSize, err := betb.getBetStreamSize(*stream)
	if err == nil {
		newBet := new(Bet)
		err = newBet.Deserialize((*stream)[:betSize])
		betb.Bets = append(betb.Bets, *newBet)
		*stream = (*stream)[betSize:]
	}
	return
}
